### PR TITLE
[hail] Memory Efficient Scan

### PR DESCRIPTION
**Pull request by danking**
cc: @lfrancioli 

We stream through the values on the leader node rather than realizing an array of all values. To do so, we serialize the RVAs to a temporary file and read them on the leader node. Leader writes out a file with each object. Workers read the file and drop objects until they reach the one of interest.

The diff is a bit unfortunate. The deleted code got split in two. The first part happens before we write the first file. The second part happens on the leader node (as it did before), but we `scanLeft` on an iterator of RVAs that we read from a series of files (each file containing one RVA).

Instead of broadcasting or serializing the result, we write it to a file and modify `itF` to read the data from a file.

We technically write one more scan value than necessary (the last one) to the second file 🤷‍♀ .

We don't have closable iterators so I had to manually track and close the input streams.

cc: @cseed

### CODE REVIEW

File: hail/src/main/scala/is/hail/expr/ir/TableIR.scala [hunk 1]
```diff
@@ -851,7 +852,71 @@ case class TableMapRows(child: TableIR, newRow: IR) extends TableIR {
...
+        new ObjectOutputStream(os).writeObject(aggs)
+        1
+      })
+
+      val hConf = hc.hadoopConf
+      val inputStreams = new Array[ObjectInputStream](partFiles.length)
+      val globallyScannedFile = HailContext.get.getTemporaryFile()
+      try {
+        hConf.writeFile(globallyScannedFile) { os =>
+          using(new ObjectOutputStream(os)) { oos =>
```
Comment (cseed) [inline hail/src/main/scala/is/hail/expr/ir/TableIR.scala]: It seems like you can get rid of `inputStreams` and open one file at a time if you do, in pseudo-code:

```
  partFiles.indices.scanLeft(scanAggs) { (a, i) => 
    val a2 = using(open part file) { ois => load ith agg from part file }
    newA = a.copy()
    newA.combOp(a2)
    newA
}
```

or something like that.

Man, that copy is unfortunate.

**Comment by cseed** [comment]
Thanks @danking!  Very nice.

File: hail/src/main/scala/is/hail/expr/ir/TableIR.scala (hunk 1 above)
Comment (danking) [reply hail/src/main/scala/is/hail/expr/ir/TableIR.scala]: ah good point

**Comment by tpoterba** [comment]
hold up -- we need to benchmark

**Comment by tpoterba** [comment]
I want to see how long it takes to do the following before/after this change, locally:

```
ht = hl.utils.range_table(1000, n_partitions=1000)
ht = ht.annotate(rank = hl.agg.count())._force_count()
```

**Comment by cseed** [comment]
If only we had, y'know, a benchmarking system that could, y'know, run benchmarks.

**Comment by danking** [comment]
I ran two tests. The one you suggested, 10k/10k, and one with a better work-to-partition ratio: 1M/1k. I was a bit surprised to see that the cost ratio is larger with fewer partitions and more work. We range from 15x to 3x slower. 

This branch:
```python
In [3]: %%time  
   ...:  
   ...: import hail as hl 
   ...: ht = hl.utils.range_table(10000, n_partitions=10000) 
   ...: ht = ht.annotate(rank = hl.scan.count())._force_count() 
CPU times: user 17 ms, sys: 6.36 ms, total: 23.4 ms
Wall time: 1min 16s

In [1]: %%time  
   ...:  
   ...: import hail as hl 
   ...: ht = hl.utils.range_table(1000000, n_partitions=1000) 
   ...: ht = ht.annotate(rank = hl.scan.count())._force_count()                                                                                                                                             
CPU times: user 1.42 s, sys: 303 ms, total: 1.73 s
Wall time: 15.2 s

```

Master `0.2.14-4da055db5a7b`
```python
In [1]: %%time  
   ...:  
   ...: import hail as hl 
   ...: ht = hl.utils.range_table(10000, n_partitions=10000) 
   ...: ht = ht.annotate(rank = hl.scan.count())._force_count()                                                                                                                                             
CPU times: user 1.45 s, sys: 333 ms, total: 1.78 s
Wall time: 24.6 s
In [3]: %%time  
   ...:  
   ...: import hail as hl 
   ...: ht = hl.utils.range_table(1000000, n_partitions=1000) 
   ...: ht = ht.annotate(rank = hl.scan.count())._force_count()                                                                                                                                             
CPU times: user 6.23 ms, sys: 1.96 ms, total: 8.19 ms
Wall time: 1.33 s
```

**Comment by tpoterba** [comment]
And this is on a laptop with an SSD, right? it'll be even worse on the cloud, I think.

Should we parameterize this behavior?

**Comment by danking** [comment]
Yes, switch to Zulip, I have an idea.

**Comment by danking** [comment]
Better idea forthcoming.

**Pull request closed without merging**